### ISSUE TITLE
MongoDB fails due to BSON document size limit!

### ISSUE THREAD

**Issue by AymenFJA** [radical-cybertools/radical.pilot#2479]
Hello all, 

I have a use case named [Colmena](https://github.com/exalearn/colmena/tree/value-server) which is based on the idea of `Thinker` and `Doer` (`Consumer`/`Producer`) that I have been working on for a while. We were able to develop the necessary requirements to push this use case forward.

This use case requires a `Radis` server to be deployed on the login node to work as a communication point between the `Thinker` and `Doer`. The use case output is a Class object with several parameters that need to be returned in a serialized way to be processed by the `Thinker Class` to generate a new `input Class`.

Apparently, `Redis` works like MongoDB in terms of shipping the tasks input/output of more than 100 MB.

I was successfully able to run this use case locally with a `small input size` using MongoDB instead of `Redis`. Unfortunately, (1) the use case output grows with the number of times that a certain function executes (target function) until it reaches the desired target! (2) and the real use case input/output is big by default (>100 MB per task) leading RADICAL-Pilot to fail to submit `Colmena tasks` ( specifically 128 MB) due to the MongoDB `BSON document size limit` which is (16 MB).  Example of the error when running the use case on `XSEDE Expanse HPC`:

```
2021-12-02 10:19:17,625 - colmena.method_server.parsl - INFO - Launched Parsl DFK. Process id: 11781
2021-12-02 10:19:17,625 - colmena.method_server.parsl - INFO - Starting the error-handler thread
2021-12-02 10:19:17,625 - colmena.method_server.base - INFO - Started method server ParslMethodServer on 11781
2021-12-02 10:19:17,625 - colmena.method_server.base - INFO - Begin pulling from task queue
2021-12-02 10:19:17,978 - colmena.method_server.parsl - INFO - Received request for target_function with topic generate
2021-12-02 10:19:17,978 - parsl.dataflow.dflow - DEBUG - Task 0 will be sent to executor RADICALExecutor
2021-12-02 10:19:17,979 - parsl.dataflow.dflow - DEBUG - Adding output dependencies
2021-12-02 10:19:17,979 - parsl.dataflow.dflow - INFO - Task 0 submitted for App wrapper, not waiting on any dependency
2021-12-02 10:19:17,979 - parsl.dataflow.dflow - DEBUG - Task 0 set to pending state with AppFuture: <AppFuture super=<AppFuture at 0x1555444d9670 state=pending>>
2021-12-02 10:19:17,979 - parsl.dataflow.memoization - DEBUG - Task 0 will not be memoized
2021-12-02 10:19:19,361 - parsl.dataflow.dflow - DEBUG - Got an exception launching task
Traceback (most recent call last):
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/parsl/dataflow/dflow.py", line 492, in launch_if_ready
    exec_fu = self.launch_task(
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/parsl/dataflow/dflow.py", line 582, in launch_task
    exec_fu = executor.submit(executable, self.tasks[task_id]['resource_specification'], *args, **kwargs)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/radical/pilot/agent/executing/parsl_rp.py", line 233, in submit
    self.tmgr.submit_tasks(task)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/radical/pilot/task_manager.py", line 870, in submit_tasks
    self._session._dbs.insert_tasks(task_docs)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/radical/pilot/db/database.py", line 460, in insert_tasks
    res = bulk.execute()
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/bulk.py", line 713, in execute
    return self.__bulk.execute(write_concern, session=None)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/bulk.py", line 533, in execute
    return self.execute_command(generator, write_concern, session)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/bulk.py", line 362, in execute_command
    client._retry_with_session(
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/mongo_client.py", line 1438, in _retry_with_session
    return self._retry_internal(retryable, func, session, bulk)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/mongo_client.py", line 1470, in _retry_internal
    return func(session, sock_info, retryable)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/bulk.py", line 356, in retryable_bulk
    self._execute_command(
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/bulk.py", line 312, in _execute_command
    result, to_send = bwc.execute(cmd, ops, client)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/message.py", line 949, in execute
    request_id, msg, to_send = self._batch_command(cmd, docs)
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/message.py", line 941, in _batch_command
    request_id, msg, to_send = _do_bulk_write_command(
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/message.py", line 1445, in _do_bulk_write_command
    return _do_batched_op_msg(
  File "/home/aymen/ve/mpi/lib/python3.9/site-packages/pymongo/message.py", line 1370, in _do_batched_op_msg
    return _batched_op_msg(
pymongo.errors.DocumentTooLarge: BSON document too large (135090825 bytes) - the connected server supports BSON document sizes up to 16777216 bytes.
```
Deploying `Redis` alongside MongoDB and redirecting the task input/output to it instead of MongoDB is an option but it requires restructuring parts of the agent and the executor. 

I did read that there is a `GridFS` feature allowing MongoDB to cut the input/output size into a set of chunks < 16 MB. But RP mongo DB and based on my understanding is not built to support large file transfer rather than shipping the tasks to the agent.

Long story short, is there any way that we can avoid sending the task output via MongoDB like via staging or something (although staging input is not valid in this case because the input is generated during the runtime of the task)?

Also, is there any way to deploy `Redis` server with RP without doing deep RP reconstruction?

**Comment by AymenFJA** [comment]
This issue is solved! We have implemented Redis capabilities in RP. RP now can detect if the task is large (precisely input/output) enough to be sent via Redis queues to the agent (based on the size in MB of the object to be sent). The developed feature ensures that it won't affect or change the task scheduling or their order. The changes are reflected in the most recent commit in this branch: https://github.com/radical-cybertools/radical.pilot/commit/aabacd92f3c9d9e0d90f77be6a1cba2902f3c76d. Closing for now.
